### PR TITLE
Change conf path XDG_CACHE_HOME to XDG_STATE_HOME

### LINKED ISSUE #9738: Consider using XDG_STATE_HOME instead of cache folder for local config
## Overview
[TIP]:  # ( DO NOT include screenshots of your actual database! )
[NOTE]: # ( Give a BRIEF summary about your problem )
Transient data such as last active database, GUI size/settings are stored in `~/.cache/keepassxc/kepassxc.ini`. This means this data is prone to being deleted (as cache directories might be cleared during logout/reboot or reside on volatile storage such as ramdisks to begin with. These settings are not caches, they should go into `~/.local/share/` or `~/.local/state/` (or rather XDG_DATA_HOME) instead.

## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )
1. Start KXC
2. Observe file created under `~/.cache/`

## Expected Behavior
[NOTE]: # ( Tell us what you expected to happen )
Use `~/.local` instead

## Actual Behavior
[NOTE]: # ( Tell us what actually happens )
Uses `~/.cache`

## Context
[NOTE]: # ( Give us any additional information you may have. )
XDG_DATA_HOME is part of the [Freedesktop basedir spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html).

[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - Version 2.7.5
Revision: 9d0537b

Qt 5.15.10
Debugging mode is disabled.

Operating system: Gentoo Linux
CPU architecture: x86_64
Kernel: linux 6.1.41-gentoo-HH

Enabled extensions:
- Browser Integration
- SSH Agent
- Secret Service Integration

Cryptographic libraries:
- Botan 3.0.0

[NOTE]: # ( Pick choices based on your environment )
Operating System: Linux
Desktop Env: XFCE
Windowing System: X11

Thanks!

### PR DESCRIPTION
**Pull request by jNullj**
Keepassxc saves application state at XDG_CACHE_HOME which can be cleared on some systems periodicly.
This is not desireable as app state like window size is not consistent when openning the app.
To avoid this, this PR commit is switching the path to XDG_STATE_HOME which is more fitting based on the [freedesktop basedir spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), this will allow to prevent state file deletion as well.

Changes made based on how qt implements QStandardPaths (very similar to internal qt code)

Also added code to migrate old config files to new location and cleanup files from old location.

Fixes #9738 

[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )

## Screenshots
[TIP]:  # ( Do not include screenshots of your actual database! )
N/A

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )

Manually tested file changes in `~/.local/state` and that migration from `~/.cache` works.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (change that adds functionality)
- ✅ Breaking change (causes existing functionality to change)
- ✅ Documentation (non-code change)

### CODE REVIEW

**Review by phoerious — CHANGES_REQUESTED** [review]
You should migrate existing settings from the old location.

File: src/core/Config.cpp [hunk 1]
```diff
@@ -472,6 +472,26 @@ void Config::init(const QString& configFileName, const QString& localConfigFileN
...
+    // Move file to correct location before continuing.
+    QString oldLocalConfigPath;
+    oldLocalConfigPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + "/keepassxc";
+    QString suffix;
+#ifdef QT_DEBUG
+    suffix = "_debug";
+#endif
+    oldLocalConfigPath += QString("/keepassxc%1.ini").arg(suffix);
+    oldLocalConfigPath = QDir::toNativeSeparators(oldLocalConfigPath);
+    if (!localConfigFileName.isEmpty() && !QFile::exists(localConfigFileName) && QFile::exists(oldLocalConfigPath)) {
```
Comment (phoerious) [inline src/core/Config.cpp]: I don't think you should check for localConfigFileName being empty. That's just something that shouldn't happen and would be an error elsewhere. I think you should just check if it exists and if not, then check the old location. The definition of oldLocalConfigPath can be moved into the smaller scope of the corresponding if statement.

File: src/core/Config.cpp [hunk 2]
```diff
@@ -512,7 +532,17 @@ QPair<QString, QString> Config::defaultConfigFiles()
 #else
     // On case-sensitive Operating Systems, force use of lowercase app directories
     configPath = QStandardPaths::writableLocation(QStandardPaths::GenericConfigLocation) + "/keepassxc";
-    localConfigPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + "/keepassxc";
+    // Qt does not support XDG_STATE_HOME yet, change this once XDG_STATE_HOME is added
+    QString xdgStateHome;
+    xdgStateHome = QFile::decodeName(qgetenv("XDG_STATE_HOME"));
```
Comment (phoerious) [inline src/core/Config.cpp]: Combine these two.

Comment (jNullj) [reply src/core/Config.cpp]: My bad, fixed with https://github.com/keepassxreboot/keepassxc/pull/9755/commits/0cd5ce425e1e26854413914b0fc4e088372e5b4e

File: src/core/Config.cpp (hunk 1 above)
Comment (jNullj) [reply src/core/Config.cpp]: I missed that, the isEmpty there is not relevant, removed it with https://github.com/keepassxreboot/keepassxc/pull/9755/commits/2eecf9892eed4c73376fc76415a2ff43979e5e03
I also understand that this is not optimal to generate oldLocalConfigPath on every startup, i changed the conditions and moved it inside the first if statement, we will only generated if the config is missing, fixed in https://github.com/keepassxreboot/keepassxc/pull/9755/commits/d77ef9d4f71759d216f2ca9e04c1db3fdb038dae

File: src/core/Config.cpp [hunk 3]
```diff
@@ -476,19 +476,21 @@ void Config::init(const QString& configFileName, const QString& localConfigFileN
     // Upgrade from previous KeePassXC version which stores its config
     // in ~/.cache on Linux instead of ~/.local/state.
     // Move file to correct location before continuing.
-    QString oldLocalConfigPath;
-    oldLocalConfigPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + "/keepassxc";
-    QString suffix;
+    if (!QFile::exists(localConfigFileName)) {
+        QString oldLocalConfigPath;
+        oldLocalConfigPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation) + "/keepassxc";
```
Comment (phoerious) [inline src/core/Config.cpp]: You should not separate declaration/definition and assignment. Rest looks good.

Comment (jNullj) [reply src/core/Config.cpp]: Fixed in https://github.com/keepassxreboot/keepassxc/pull/9755/commits/85a2b740ffd2423a30d90dd9e31e2d77ac3f54aa

**Review by phoerious — APPROVED** [review]

**Comment by phoerious** [comment]
Looks good now. Thanks for your patience.

**Comment by jehoshua7** [comment]
I'm running version 2.7.10 on Kubuntu, but can't see where, in the settings, an option to change the configuration path. Having it in /.cache is not a good idea, as that path can be deleted.

**Comment by droidmonkey** [comment]
https://keepassxc.org/docs/KeePassXC_UserGuide#_environment_variables

**Comment by jehoshua7** [comment]
@droidmonkey - Thanks. I was about to set the environment variable and then did this:

`locate -i keepass`

> /home/********/.cache/keepassxc
/home/********/.cache/keepassxc/keepassxc.ini
/home/********/.config/KeePassXCrc
/home/********/.config/keepassxc
/home/********/.config/keepassxc/keepassxc.ini

and opening both the /keepassxc.ini files, they are quite different. So, if I set the environment variable to **NOT** be in the /.cache path, I run into the problem of 2 files with the same name ?  Obviously make a sub path or similar under /home/*******/.config

There are two variables , KPXC_CONFIG and KPXC_CONFIG_LOCAL ; I assume only the latter one needs changing, as I have no idea what the 'roaming' concept is.

**Comment by droidmonkey** [comment]
Just the local one, the env var includes the file name you want so set it to keepassxc_local.ini for example.

**Comment by jehoshua7** [comment]
@droidmonkey okay thanks. I see with Kubuntu, i can just add a localvariable in the menu editor for KeepassXC, so that's easier than messing around with .basrc , .profile and other files.

Related post at https://andreas.scherbaum.la/post/2024-10-03_keepassxc-bleachbit-and-a-vanishing-configuration-file/ ; yes, having the file in $HOME/.cache is not a good idea.

**Pull request merged**